File: graph.r
# install.packages("plotly")

library(plotly)

data <- read.csv("/home/conmarap/Projects/website-fingerprinting/example-fingerprints/fingerprints.csv")
colors <- c('#4AC6B7', '#1972A4', '#965F8A', '#FF7070', '#C61951', '#FAEF44', '#DEAF69')

p <- plot_ly(data, x = ~total_number_packets, y = ~no_incoming_packets, z = ~no_outgoing_packets, color = ~domain, size = ~total_incoming_sizes, colors = colors,
             marker = list(symbol = 'circle', sizemode = 'diameter'), sizes = c(5, 150),
             text = ~paste('Domain:', domain, '<br># of packets:', total_number_packets)) %>%
  layout(title = 'Website Fingerprinting',
         scene = list(xaxis = list(title = 'Total Number of packets',
                      gridcolor = 'rgb(255, 255, 255)',
                      type = 'log',
                      zerolinewidth = 1,
                      ticklen = 5,
                      gridwidth = 2),
               yaxis = list(title = 'Number of Incoming Packets',
                      gridcolor = 'rgb(255, 255, 255)',
                      zerolinewidth = 1,
                      ticklen = 5,
                      gridwith = 2),
               zaxis = list(title = 'Number of Outgoing Packets',
                            gridcolor = 'rgb(255, 255, 255)',
                            type = 'log',
                            zerolinewidth = 1,
                            ticklen = 5,
                            gridwith = 2)),
         paper_bgcolor = 'rgb(243, 243, 243)',
         plot_bgcolor = 'rgb(243, 243, 243)')

p

File: graph.py
import chart_studio.plotly as py
import plotly.graph_objects as go
import plotly.io as pio
import numpy as np
import pandas as pd

# Read the data.
data = pd.read_csv('./example-fingerprints/fingerprints.csv')

d = data['domain']
#x = data['total_number_packets']
#y = data['no_incoming_packets']
#z = data['no_outgoing_packets']
x = data['total_incoming_sizes']
y = data['ratio_incoming_to_outgoing']
z = data['total_number_packets']

xt = []
yt = []
zt = []
data = []

for i in range(0, len(x)):
    if i == len(x) - 1 or d[i] != d[i + 1]:
        data.append(go.Scatter3d(x=xt, y=yt, z=zt, mode='markers', \
            marker=dict(), name=d[i]))

        xt = []
        yt = []
        zt = []
    else:
        xt.append(x[i])
        yt.append(y[i])
        zt.append(z[i])

layout = go.Layout(margin=dict(l=0, r=0, b=0, t=0))

fig = go.Figure(data=data, layout=layout)
pio.write_html(fig, file='graphs/graph.html', auto_open=True)

